File: app/src/main/java/com/kiwilss/xview/ui/view/recyclerview/decoration/ItemDivider.java
package com.kiwilss.xview.ui.view.recyclerview.decoration;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.View;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class ItemDivider extends RecyclerView.ItemDecoration {

      private int dividerWith = 4;
      private Paint paint;
      private RecyclerView.LayoutManager layoutManager;

      // 构造方法,可以在这里做一些初始化,比如指定画笔颜色什么的
      public ItemDivider() {  
          initPaint();
          paint.setColor(0xffE6E6E6);
      }   

      private void initPaint() {    
          if (paint == null) {        
             paint = new Paint(Paint.ANTI_ALIAS_FLAG);        
             paint.setStyle(Paint.Style.FILL);    
          }
      }

      public ItemDivider setDividerWith(int dividerWith) {    
         this.dividerWith = dividerWith;   
         return this;
      }

      public ItemDivider setDividerColor(int color) {    
          initPaint();    
          paint.setColor(color);    
          return this;
       } 

      /**     
       * 指定item之间的间距(就是指定分割线的宽度)   回调顺序 1     
       * @param outRect Rect to receive the output.      
       * @param view    The child view to decorate     
       * @param parent  RecyclerView this ItemDecoration is decorating     
       * @param state   The current state of RecyclerView.     
       */    
       @Override    
       public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
           super.getItemOffsets(outRect, view, parent, state);    
            if (layoutManager == null) {    
               layoutManager = parent.getLayoutManager();
            }
            // 适用 LinearLayoutManager 和 GridLayoutManager
            if (layoutManager instanceof LinearLayoutManager) {
               int orientation = ((LinearLayoutManager) layoutManager).getOrientation();
               if (orientation == LinearLayoutManager.VERTICAL) {
                   // 水平分割线将绘制在item底部        
                   outRect.bottom = dividerWith;    
               } else if (orientation == LinearLayoutManager.HORIZONTAL) {
                   // 垂直分割线将绘制在item右侧        
                   outRect.right = dividerWith;   
               }    
               if (layoutManager instanceof GridLayoutManager) {
                   GridLayoutManager.LayoutParams lp = (GridLayoutManager.LayoutParams) view.getLayoutParams();
                   // 如果是 GridLayoutManager 则需要绘制另一个方向上的分割线       
                   if (orientation == LinearLayoutManager.VERTICAL && lp != null && lp.getSpanIndex() > 0) {
                      // 如果列表是垂直方向,则最左边的一列略过            
                      outRect.left = dividerWith;
                   } else if (orientation == LinearLayoutManager.HORIZONTAL && lp != null && lp.getSpanIndex() > 0) {
                      // 如果列表是水平方向,则最上边的一列略过            
                      outRect.top = dividerWith;
                   }    
               }
           }  
       }    

       /**     
        * 在item 绘制之前调用(就是绘制在 item 的底层)  回调顺序 2     
        * 一般分割线在这里绘制     
        * 看到canvas,对自定义控件有一定了解的话,就能想到为什么说给RecyclerView设置分割线更灵活了
        * @param c      Canvas to draw into     
        * @param parent RecyclerView this ItemDecoration is drawing into     
        * @param state  The current state of RecyclerView     
        */    
        @Override    
        public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
            super.onDraw(c, parent, state);
            RecyclerView.LayoutManager layoutManager = parent.getLayoutManager();
            int spanCount = 0;
            if (layoutManager instanceof GridLayoutManager){
               spanCount =  ((GridLayoutManager) layoutManager).getSpanCount();
            }
            int childCount = parent.getChildCount();
            int count = childCount % spanCount;
            if (count == 0){
                count = spanCount;
            }
            // 这个值是为了补偿横竖方向上分割线交叉处间隙
            int offSet = (int) Math.ceil(dividerWith * 1f / 2);
            for (int i = 0; i < parent.getChildCount(); i++) {    
                View child = parent.getChildAt(i);    
                RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
                int left1 = child.getRight() + params.rightMargin;
                int right1 = left1 + dividerWith;
                int top1 = child.getTop() - offSet - params.topMargin;
                int bottom1 = child.getBottom() + offSet + params.bottomMargin;
                //绘制分割线(矩形)
                c.drawRect(left1, top1, right1, bottom1, paint);
                int left2 = child.getLeft() - offSet - params.leftMargin;
                int right2 = child.getRight() + offSet + params.rightMargin;
                int top2 = child.getBottom() + params.bottomMargin;
                int bottom2 = top2 + dividerWith;
                //绘制分割线(矩形)
                //c.drawRect(left2, top2, right2, bottom2, paint);
                //判断最后一行不画
                if (i < parent.getChildCount() -count){
                    c.drawRect(left2, top2, right2, bottom2, paint);
                }
             }         
        }    

        /**     
         * 在item 绘制之后调用(就是绘制在 item 的上层)  回调顺序 3     
         * 也可以在这里绘制分割线,和上面的方法 二选一     
         * @param c      Canvas to draw into     
         * @param parent RecyclerView this ItemDecoration is drawing into     
         * @param state  The current state of RecyclerView     
         */    
         @Override   
         public void onDrawOver(Canvas c, RecyclerView parent, RecyclerView.State state) {
            super.onDrawOver(c, parent, state);    
         }
     }

File: app/src/main/java/com/kiwilss/xview/ui/view/recyclerview/decoration/ItemDivider.kt
/**
 * Copyright (C), 2017-2020, XXX有限公司
 * FileName: ItemDivider
 * Author:   Administrator
 * Date:     2020/8/1 16:02
 * Description: {DESCRIPTION}
 * History:
 * <author>          <time>          <version>          <desc>
 * 作者姓名           修改时间           版本号              描述
 */

package com.kiwilss.xview.ui.view.recyclerview.decoration

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.view.View
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.kiwilss.xview.R
import com.kiwilss.xview.utils.LogUtils
import kotlin.math.ceil

/**
 *@FileName: ItemDivider
 *@author : Lss Administrator
 * @e-mail : kiwilss@163.com
 * @time   : 2020/8/1
 * @desc   : {DESCRIPTION}
 */
class ItemDivider(
    private val firstTop: Boolean = false, private val lastBottom: Boolean = false,
    private val left: Boolean = false, private val right: Boolean = false,
    private val dividerWidth: Int = 2, val color: Int = Color.parseColor("#f2f2f2")
                    ): RecyclerView.ItemDecoration() {

    private var layoutManager: RecyclerView.LayoutManager? = null
    private var paint: Paint? = null

    init {
        //初始化paint
        if (paint == null) {
            paint = Paint(Paint.ANTI_ALIAS_FLAG)
            paint!!.style = Paint.Style.FILL
            paint!!.color = color
        }
    }
    override fun getItemOffsets(
        outRect: Rect,
        view: View,
        parent: RecyclerView,
        state: RecyclerView.State
    ) {
        super.getItemOffsets(outRect, view, parent, state)
        if (layoutManager == null) {
            layoutManager = parent.layoutManager
        }
        if (layoutManager is LinearLayoutManager) {//线性
            LogUtils.e("线性-->")
            val orientation = (layoutManager as LinearLayoutManager).orientation
            if (orientation == LinearLayoutManager.VERTICAL) {
                //垂直方向用的较多,特殊处理第一个Item顶部是否绘制分割线,最后一个Item底部是否绘制分割线
                //中间的Item全都绘制在底部
                if (parent.getChildLayoutPosition(view) == 0) {
                    if (firstTop){
                        outRect.top = dividerWidth
                    }
                    LogUtils.e(firstTop,"firstTop")
                    outRect.bottom = dividerWidth
                }else if (parent.getChildLayoutPosition(view) + 1 == parent.adapter?.itemCount){
                    if (lastBottom){
                        outRect.bottom = dividerWidth
                    }
                }else{
                    LogUtils.e("bottom--->linear")
                    // 水平分割线将绘制在item底部
                    outRect.bottom = dividerWidth
                }
            } else if (orientation == LinearLayoutManager.HORIZONTAL) {
                //水平方向,top代表第一个item左侧是否绘制,bottom代表最后一个item右边是否绘制
                if (parent.getChildLayoutPosition(view) == 0) {
                    if (firstTop){
                        outRect.left = dividerWidth
                    }
                }else if (parent.getChildLayoutPosition(view) + 1 == parent.adapter?.itemCount){
                    if (lastBottom){
                        outRect.right = dividerWidth
                    }
                }else{
                    // 垂直分割线将绘制在item右侧
                    outRect.right = dividerWidth
                }
            }
            if (layoutManager is GridLayoutManager){//网格
                //处理一下特殊情况,firstTop第一行是否绘制,lastBottom最后一行是否绘制,left左侧是否绘制,right右侧是否绘制(默认全都不绘制)
                val childLayoutPosition = parent.getChildLayoutPosition(view)
                val childCount = (layoutManager as GridLayoutManager).childCount
                val spanCount = (layoutManager as GridLayoutManager).spanCount
                val lp = view.layoutParams as GridLayoutManager.LayoutParams
                LogUtils.e("$childLayoutPosition + ---------spancount= == $spanCount---------count=$childCount")
                if (orientation == LinearLayoutManager.VERTICAL){
                    //如果只有一行时,特殊情况处理
                    if (childCount <= spanCount){
                        if (firstTop){
                            outRect.top = dividerWidth
                        }
                        if (lastBottom){
                            outRect.bottom = dividerWidth
                        }
                        if (left && lp.spanIndex == 0){//第一个的左边绘制
                            outRect.left = dividerWidth
                        }
                        if (right && lp.spanIndex == spanCount - 1){//最后一个右边绘制
                            outRect.right = dividerWidth
                        }
                        //中间的全都绘制左边
                        if ( lp.spanIndex > 0) {
                            // 如果列表是垂直方向,则最左边的一列略过
                            outRect.left = dividerWidth
                        }
                    }


                }else if (orientation == LinearLayoutManager.HORIZONTAL && lp.spanIndex > 0) {
                    // 如果列表是水平方向,则最上边的一列略过
                    outRect.top = dividerWidth
                }


            }
        }
    }

    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDraw(c, parent, state)
//        val layoutManager = parent.layoutManager
//        var spanCount = 0
//        if (layoutManager is GridLayoutManager) {
//            spanCount = layoutManager.spanCount
//        }
//        val childCount = parent.childCount
//        var count = childCount % spanCount
//        if (count == 0) {
//            count = spanCount
//        }
        // 这个值是为了补偿横竖方向上分割线交叉处间隙
        // 这个值是为了补偿横竖方向上分割线交叉处间隙
        val offSet = ceil(dividerWidth * 1f / 2.toDouble()).toInt()
        for (i in 0 until parent.childCount) {
            val child = parent.getChildAt(i)
            val params =
                child.layoutParams as RecyclerView.LayoutParams
            val left1 = child.right + params.rightMargin
            val right1: Int = left1 + dividerWidth
            val top1 = child.top - offSet - params.topMargin
            val bottom1 = child.bottom + offSet + params.bottomMargin
            //绘制分割线(绘制竖直方向的矩形)
            c.drawRect(
                left1.toFloat(), top1.toFloat(), right1.toFloat(), bottom1.toFloat(),
                paint!!
            )
            val left2 = child.left - offSet - params.leftMargin
            val right2 = child.right + offSet + params.rightMargin
            val top2 = child.bottom + params.bottomMargin
            val bottom2: Int = top2 + dividerWidth
            //绘制分割线(绘制水平方向的矩形)
            c.drawRect(left2.toFloat(), top2.toFloat(), right2.toFloat(), bottom2.toFloat(),
                paint!!
            )
            //判断最后一行不画
//            if (i < parent.childCount - count) {
//                c.drawRect(
//                    left2.toFloat(),
//                    top2.toFloat(),
//                    right2.toFloat(),
//                    bottom2.toFloat(),
//                    paint!!
//                )
//            }
        }
    }

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDrawOver(c, parent, state)
    }
}